vms/platformvm/state/state.go

```diff
@@ -1712,22 +1853,75 @@ func (s *state) loadPendingValidators() error {
 	)
 }
 
-// Invariant: initValidatorSets requires loadCurrentValidators to have already
-// been called.
+// Invariant: initValidatorSets requires loadActiveSubnetOnlyValidators and
+// loadCurrentValidators to have already been called.
 func (s *state) initValidatorSets() error {
-	for subnetID, validators := range s.currentStakers.validators {
+	// Load ACP77 validators
+	for validationID, sov := range s.activeSOVLookup {
+		pk := bls.PublicKeyFromValidUncompressedBytes(sov.PublicKey)
+		if err := s.validators.AddStaker(sov.SubnetID, sov.NodeID, pk, validationID, sov.Weight); err != nil {
+			return err
+		}
+	}
+
+	// Load inactive weights
+	it := s.weightsDB.NewIterator()
+	defer it.Release()
+
+	for it.Next() {
+		subnetID, err := ids.ToID(it.Key())
+		if err != nil {
+			return err
+		}
+
+		totalWeight, err := database.ParseUInt64(it.Value())
+		if err != nil {
+			return err
+		}
+
+		activeWeight, err := s.validators.TotalWeight(subnetID)
+		if err != nil {
+			return err
+		}
+
+		inactiveWeight, err := safemath.Sub(totalWeight, activeWeight)
+		if err != nil {
+			// This should never happen, as the total weight should always be at
+			// least the sum of the active weights.
+			return err
+		}
+		if inactiveWeight == 0 {
+			continue
+		}
+
+		if err := s.validators.AddStaker(subnetID, ids.EmptyNodeID, nil, ids.Empty, inactiveWeight); err != nil {
```

Comment:
shall we use a special ID (like 111, 001 etc) for inactive validators (rather than empty nodeID)

Reply:
Because only the `EmptyNodeID` is prevented from being issued to the network, I don't think we have any other options than using the `EmptyNodeID` here. Unless we modified the validator manager to know about inactive weight explicitly.

Reply:
not sure I follow, doesn't the validator manager already know abut total inactive weights (by `EmptyNodeID`)?

Reply:
The validator manager does not treat `ids.EmptyNodeID` differently than other `NodeID`s.

Reply:
My only concern and question is would it be safe against the proposer vm's special casing of emptyNodeID? I guess we still need to explicitly handle it  in proposer VM even if we decide to use another ID

Reply:
> My only concern and question is would it be safe against the proposer vm's special casing of emptyNodeID

I have a local branch working to remove that special casing. Not quite ready to push up yet though